core/grant/service.go

```diff
@@ -257,3 +270,176 @@ func (s *Service) expiredInActiveUserAccess(ctx context.Context, timeLimiter cha
 		}
 	}
 }
+
+type ImportAccessCriteria struct {
+	ProviderID   string `validate:"required"`
+	ResourceIDs  []string
+	ResouceTypes []string
+	ResourceURNs []string
+}
+
+func (s *Service) ImportAccess(ctx context.Context, criteria ImportAccessCriteria) ([]*domain.Grant, error) {
+	p, err := s.providerService.GetByID(ctx, criteria.ProviderID)
+	if err != nil {
+		return nil, fmt.Errorf("getting provider details: %w", err)
+	}
+
+	listResourcesFilter := domain.ListResourcesFilter{
+		ProviderType: p.Type,
+		ProviderURN:  p.URN,
+	}
+	if criteria.ResourceIDs != nil {
+		listResourcesFilter.IDs = criteria.ResourceIDs
+	} else {
+		listResourcesFilter.ResourceTypes = criteria.ResouceTypes
+		listResourcesFilter.ResourceURNs = criteria.ResourceURNs
+	}
+	resources, err := s.resourceService.Find(ctx, listResourcesFilter)
+	if err != nil {
+		return nil, fmt.Errorf("getting resources: %w", err)
+	}
+
+	_, resourceAccess, err := s.providerService.ListAccess(ctx, criteria.ProviderID, resources)
```

Comment:
This `ListAccess` in provider service has 1 DB call to get provider `GetByID`. I don't think it is necessary because we have called `GetByID` earlier in this function.
I think we can have a `ListAccess` that does not depend on DB call, maybe try to just have a `GetClient` exposed or having a `ListAccess` function that accept provider type argument instead of ID.

Reply:
changed `provider.ListAccess` to receive `provider` struct, however it's still need to be part of provider service so it can be used by the other service (in this case grant service)